{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEE2 STAR counts\n",
    "full processing from dee2 to h5 gse x gene file. symbol in genes. gene_sybol in genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "make gene index...\n",
      "gene quantity: 4497\n",
      "read meta file...\n",
      "real SRR quantity: 8632\n",
      "SRRs with gse: 4571\n",
      "make SRR and gsm indexes...\n",
      "GSMs quantity:\n",
      "start grouping...\n",
      "groupped\n",
      "reordered\n",
      "start h5...\n",
      "create file and write meta...\n",
      "start process expression...\n",
      "#\n",
      "0.11584800741427248\n",
      "#\n",
      "0.23169601482854496\n",
      "#\n",
      "0.34754402224281744\n",
      "#\n",
      "0.4633920296570899\n",
      "#\n",
      "0.5792400370713624\n",
      "#\n",
      "0.6950880444856349\n",
      "#\n",
      "0.8109360518999074\n",
      "#\n",
      "0.9267840593141798\n",
      "#\n",
      "1\n",
      "Wall time: 1min 1s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import datetime\n",
    "import bz2\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.sparse as sps\n",
    "from scipy.sparse import isspmatrix_csr,csr_matrix,csc_matrix,issparse,isspmatrix_csr\n",
    "import math\n",
    "cache_dir=\"G://datasets/DEE2/\"\n",
    "#cache_dir=\"./\"\n",
    "organism_name= 'ecoli'#'hsapiens' 'ecoli' 'mmusculus' 'athaliana' 'drerio' 'scerevisiae'\n",
    "metadata_file = cache_dir+\"metadata/\"+ organism_name+\"_metadata_20210225.tsv\"\n",
    "data_file_name= cache_dir+\"raw_files/star_counts/\"+organism_name+\"_se_20210224.tsv.bz2\" #celegans_se_20200530.tsv.bz2 dmelanogaster_se_20200530.tsv.bz2 ecoli_se_20200530.tsv.bz2\n",
    "#drerio_se_20200530.tsv.bz2\n",
    "#hsapiens_se_20200528.tsv.bz2\n",
    "h5_gsm_name=cache_dir+\"star_h5/\"+organism_name+ \"_star_matrix_20210224.h5\"\n",
    "#h5_gsm_name = \"/scratch2/mkleverov/dee2/star_h5/\"+organism_name+ \"_star_matrix.h5\"\n",
    "gene_info_file_name =cache_dir+\"GeneInfo/\"+organism_name+\"_GeneInfo.tsv\"\n",
    "dt = h5py.special_dtype(vlen=np.str)\n",
    "now = datetime.datetime.now()\n",
    "print(\"make gene index...\")\n",
    "gene_info=pd.read_csv(filepath_or_buffer = gene_info_file_name, sep=\"\\t\")\n",
    "gene_ind= {gene:ind for ind,gene in enumerate(gene_info[\"GeneID\"])}\n",
    "n_genes=len(gene_ind)\n",
    "print(\"gene quantity: \"+str(n_genes))\n",
    "print(\"read meta file...\")\n",
    "iter_meta = pd.read_csv(filepath_or_buffer=metadata_file,sep=\"\\t\", iterator=True, chunksize=1000)\n",
    "meta_df = pd.concat([chunk for chunk in iter_meta])\n",
    "geo_acc_name = meta_df.columns[5]#'GSE_accession' # may be \"GSE_accession\" or \"Sample_name\"\n",
    "geo_gse_name = meta_df.columns[6]#'Sample_name' # may be \"GEO_series\" \"SampleName\" or \"Sample_name\"(ecoli)\n",
    "meta_df[geo_gse_name]= meta_df[geo_gse_name].astype(str)\n",
    "true_srr=len(meta_df)\n",
    "print(\"real SRR quantity: \" +str(true_srr))\n",
    "meta_df=meta_df[meta_df[geo_acc_name].str.startswith('GSM')]\n",
    "n_srr= len(meta_df)\n",
    "print(\"SRRs with gse: \"+str(n_srr))\n",
    "print(\"make SRR and gsm indexes...\")\n",
    "srr_to_gsm= dict(zip(meta_df[\"SRR_accession\"],meta_df[geo_acc_name]))\n",
    "srr_ind= {srr:ind for ind,srr in enumerate(srr_to_gsm.keys())}\n",
    "gsm_ind= {gsm:ind for ind,gsm in enumerate(pd.unique(meta_df[geo_acc_name]))}\n",
    "n_gsm= len(gsm_ind)\n",
    "print(\"GSMs quantity:\" )\n",
    "gsm_from_srr_matrix= sps.lil_matrix((n_gsm,n_srr))\n",
    "for srr_key in srr_to_gsm:\n",
    "    gsm_from_srr_matrix[gsm_ind[srr_to_gsm[srr_key]],srr_ind[srr_key]]=1\n",
    "print(\"start grouping...\")\n",
    "col_list = [geo_acc_name,\"SRR_accession\",\"Model\",\"LoadDate\",\"LibrarySelection\",\"LibrarySource\",\"LibraryStrategy\",\"ScientificName\",geo_gse_name,\"Consent\",\"ReleaseDate\",\"QC_summary\",\"BioSample\"]\n",
    "meta_df=meta_df[col_list]\n",
    "meta_df =meta_df.groupby(geo_acc_name).agg([lambda x: ';'.join(x)])\n",
    "print(\"groupped\")\n",
    "meta_df=meta_df.loc[gsm_ind.keys()]\n",
    "print(\"reordered\")\n",
    "print(\"start h5...\")\n",
    "with h5py.File(h5_gsm_name, 'w') as h5_gse:\n",
    "    print(\"create file and write meta...\")\n",
    "    h5_gse.create_dataset('/meta/info/version', data=\"9\")\n",
    "    h5_gse.create_dataset('/meta/info/creation_date', data=now.strftime(\"%Y.%m.%d\"))\n",
    "    dt = h5py.special_dtype(vlen=np.str)\n",
    "    h5_gse.create_dataset('/meta/genes/ensembl_gene_id',data=np.array(list(gene_ind.keys()),dtype=dt),chunks =(len(gene_ind),),compression=\"gzip\", compression_opts=7)\n",
    "    h5_gse.create_dataset('/meta/genes/genes',data=np.array(gene_info[\"GeneSymbol\"],dtype=dt),chunks =(len(gene_ind),),compression=\"gzip\", compression_opts=7)\n",
    "    h5_gse.create_dataset('/meta/genes/gene_symbol',data=np.array(gene_info[\"GeneSymbol\"],dtype=dt),chunks =(len(gene_ind),),compression=\"gzip\", compression_opts=7)\n",
    "    h5_gse.create_dataset('/meta/info/description', data=\"Contain all SRRs from dee2 for which sample_geo_accession starts with gse. Star estimates was groupped by sample_geo_accession and summed up\")\n",
    "    h5_gse.create_dataset('/meta/samples/geo_accession',data=np.array(list(gsm_ind.keys()), dtype =dt),chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/SRR_accession',data= np.reshape(np.array(meta_df[\"SRR_accession\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/instrument_model',data=np.reshape(np.array(meta_df[\"Model\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/last_update_date',data=np.reshape(np.array(meta_df[\"LoadDate\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/library_selection',data=np.reshape(np.array(meta_df[\"LibrarySelection\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/library_source',data=np.reshape(np.array(meta_df[\"LibrarySource\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/library_strategy',data=np.reshape(np.array(meta_df[\"LibraryStrategy\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/organism_ch1',data=np.reshape(np.array(meta_df[\"ScientificName\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/series_id',data=np.reshape(np.array(meta_df[geo_gse_name]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/status',data=np.reshape(np.array(meta_df[\"Consent\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/submission_date',data=np.reshape(np.array(meta_df[\"ReleaseDate\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/quality',data=np.reshape(np.array(meta_df[\"QC_summary\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/title',data=np.reshape(np.array(meta_df[\"BioSample\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/type',data=np.full(len(meta_df[\"BioSample\"]),\"SRA\", dtype=dt),chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    exp_data=h5_gse.create_dataset(\"/data/expression\", (n_genes,n_gsm),dtype= 'i4',chunks =(200,200), compression=\"gzip\", compression_opts=7)\n",
    "    srr_per_time=1000\n",
    "    iter_data = pd.read_csv(filepath_or_buffer=data_file_name,sep=\"\\t\", iterator=True, chunksize=srr_per_time*n_genes,names=[\"srr\",\"gene\",\"se\"])\n",
    "    proc_srr= 0\n",
    "    gsm_from_srr_matrix=csr_matrix(gsm_from_srr_matrix)\n",
    "    print(\"start process expression...\")\n",
    "    for chunk in iter_data:\n",
    "        global_ids =[]\n",
    "        local_ids=[]\n",
    "        local_srr_ind= {srr:ind for ind,srr in enumerate(pd.unique(chunk['srr']))}\n",
    "        srr_count=len(local_srr_ind)\n",
    "        proc_srr=proc_srr+srr_count\n",
    "        for cur_srr in local_srr_ind:\n",
    "            try:\n",
    "                srr_id= srr_ind[cur_srr]\n",
    "                global_ids.append(srr_id)\n",
    "                local_ids.append(local_srr_ind[cur_srr])\n",
    "            except(KeyError):\n",
    "                continue\n",
    "        if local_ids:\n",
    "            print(\"#\")\n",
    "            raw_matrix= chunk[\"se\"].values.astype(int).reshape(srr_count,n_genes)\n",
    "            A=gsm_from_srr_matrix[:,global_ids] #size = all_gsm x needed_srr\n",
    "            gsm_mask= np.unique(A.nonzero()[0]) # find needed gsm (rows in matrix above)\n",
    "            A = A.tocsr() \n",
    "            B=csr_matrix(raw_matrix[local_ids,])\n",
    "            A = A.dot(B)\n",
    "            A=A.transpose()\n",
    "            exp_data[0:n_genes,gsm_mask]+= A[0:n_genes,gsm_mask]            \n",
    "        if proc_srr%500 ==0:\n",
    "            print(proc_srr/true_srr)       \n",
    "    h5_gse.flush()\n",
    "print(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEE2 STAR counts/pandas\n",
    "\n",
    "Same as above, but more with pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "make gene index...\n",
      "gene quantity: 32883\n",
      "read meta file...\n",
      "real SRR quantity: 28753\n",
      "SRRs with gse: 19677\n",
      "make SRR and gsm indexes...\n",
      "GSMs quantity:15773\n",
      "start h5...\n",
      "create file and write info...\n",
      "start process expression...\n",
      "0.03477897958473898\n",
      "0.06955795916947796\n",
      "0.10433693875421696\n",
      "0.13911591833895592\n",
      "0.17389489792369492\n",
      "0.2086738775084339\n",
      "0.24345285709317288\n",
      "0.27823183667791185\n",
      "0.31301081626265087\n",
      "0.34778979584738984\n",
      "0.3825687754321288\n",
      "0.4173477550168678\n",
      "0.4521267346016068\n",
      "0.48690571418634576\n",
      "0.5216846937710847\n",
      "0.5564636733558237\n",
      "0.5912426529405628\n",
      "0.6260216325253017\n",
      "0.6608006121100407\n",
      "0.6955795916947797\n",
      "0.7303585712795186\n",
      "0.7651375508642576\n",
      "0.7999165304489966\n",
      "0.8346955100337357\n",
      "0.8694744896184746\n",
      "0.9042534692032136\n",
      "0.9390324487879526\n",
      "0.9738114283726915\n",
      "publish meta...\n",
      "start grouping...\n",
      "groupped\n",
      "reordered\n",
      "1\n",
      "Wall time: 1h 7min 48s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import datetime\n",
    "import bz2\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.sparse as sps\n",
    "from scipy.sparse import isspmatrix_csr,csr_matrix,csc_matrix,issparse,isspmatrix_csr, lil_matrix\n",
    "import math\n",
    "cache_dir=\"G://datasets/DEE2/\"\n",
    "#cache_dir=\"./\"\n",
    "organism_name= 'rnorvegicus'#'hsapiens' 'ecoli' 'mmusculus' 'athaliana' 'drerio' 'scerevisiae'celegans dmelanogaster  rnorvegicus\n",
    "metadata_date = \"20211102\"\n",
    "metadata_file = cache_dir+\"metadata/\"+ organism_name+\"_metadata_\"+metadata_date+\".tsv\"\n",
    "source_date = \"20211103\"\n",
    "source_name= organism_name+\"_se_\"+source_date+\".tsv.bz2\"\n",
    "data_file_name= cache_dir+\"raw_files/star_counts/\"+source_name #celegans_se_20200530.tsv.bz2 dmelanogaster_se_20200530.tsv.bz2 ecoli_se_20200530.tsv.bz2\n",
    "#drerio_se_20200530.tsv.bz2\n",
    "#hsapiens_se_20200528.tsv.bz2\n",
    "h5_gsm_name=cache_dir+\"star_h5/\"+organism_name+ \"_star_matrix_\"+source_date+\".h5\"\n",
    "#h5_gsm_name = \"/scratch2/mkleverov/dee2/star_h5/\"+organism_name+ \"_star_matrix.h5\"\n",
    "gene_info_file_name =cache_dir+\"GeneInfo/\"+organism_name+\"_GeneInfo.tsv\"\n",
    "dt = h5py.special_dtype(vlen=np.str)\n",
    "now = datetime.datetime.now()\n",
    "print(\"make gene index...\")\n",
    "gene_info=pd.read_csv(filepath_or_buffer=gene_info_file_name,sep=\"\\t\")\n",
    "gene_ind= {gene:ind for ind,gene in enumerate(gene_info[\"GeneID\"])}\n",
    "n_genes=len(gene_ind)\n",
    "print(\"gene quantity: \"+str(n_genes))\n",
    "print(\"read meta file...\")\n",
    "iter_meta = pd.read_csv(filepath_or_buffer=metadata_file,sep=\"\\t\", iterator=True, chunksize=1000)\n",
    "meta_df = pd.concat([chunk for chunk in iter_meta])\n",
    "geo_acc_name = meta_df.columns[5]#'GSE_accession' # may be \"GSE_accession\" or \"Sample_name\"\n",
    "geo_gse_name = meta_df.columns[6]#'Sample_name' # may be \"GEO_series\" \"SampleName\" or \"Sample_name\"(ecoli)\n",
    "meta_df[geo_gse_name]= meta_df[geo_gse_name].astype(str)\n",
    "true_srr=len(meta_df)\n",
    "print(\"real SRR quantity: \" +str(true_srr))\n",
    "meta_df=meta_df[meta_df[geo_acc_name].str.startswith('GSM')]\n",
    "n_srr= len(meta_df)\n",
    "print(\"SRRs with gse: \"+str(n_srr))\n",
    "print(\"make SRR and gsm indexes...\")\n",
    "srr_to_gsm= dict(zip(meta_df[\"SRR_accession\"],meta_df[geo_acc_name]))\n",
    "srr_ind= {srr:ind for ind,srr in enumerate(srr_to_gsm.keys())}\n",
    "gsm_ind= {gsm:ind for ind,gsm in enumerate(pd.unique(meta_df[geo_acc_name]))}\n",
    "col_list = [geo_acc_name,\"SRR_accession\",\"Model\",\"LoadDate\",\"LibrarySelection\",\"LibrarySource\",\"LibraryStrategy\",\"ScientificName\",geo_gse_name,\"Consent\",\"ReleaseDate\",\"QC_summary\",\"BioSample\"]\n",
    "meta_df=meta_df[col_list]\n",
    "n_gsm= len(gsm_ind)\n",
    "print(\"GSMs quantity:\" + str(n_gsm) )\n",
    "gsm_from_srr_matrix= sps.lil_matrix((n_gsm,n_srr))\n",
    "for srr_key in srr_to_gsm:\n",
    "    gsm_from_srr_matrix[gsm_ind[srr_to_gsm[srr_key]],srr_ind[srr_key]]=1\n",
    "print(\"start h5...\")\n",
    "with h5py.File(h5_gsm_name, 'w') as h5_gse:\n",
    "    print(\"create file and write info...\")\n",
    "    h5_gse.create_dataset('/meta/info/version', data=\"10\")\n",
    "    h5_gse.create_dataset('/meta/info/creation_date', data=now.strftime(\"%Y-%m-%d\"))\n",
    "    h5_gse.create_dataset('/meta/info/database', data=\"dee2\")\n",
    "    h5_gse.create_dataset('/meta/info/source_file', data=source_name)\n",
    "    h5_gse.create_dataset('/meta/info/author', data=\"Maksim Kleverov\")\n",
    "    h5_gse.create_dataset('/meta/info/contact', data=\"klevermx@gmail.com\")\n",
    "    exp_data=h5_gse.create_dataset(\"/data/expression\", (n_genes,n_gsm),dtype= 'i4',chunks =(200,200), compression=\"gzip\", compression_opts=7)\n",
    "    srr_per_time=200\n",
    "    iter_data = pd.read_csv(filepath_or_buffer=data_file_name,sep=\"\\t\", iterator=True, chunksize=srr_per_time*n_genes,names=[\"srr\",\"gene\",\"se\"])\n",
    "    proc_srr= 0\n",
    "    gsm_from_srr_matrix=csr_matrix(gsm_from_srr_matrix)\n",
    "    print(\"start process expression...\")\n",
    "    bad_srrs = []\n",
    "    for chunk in iter_data:\n",
    "        count_genes = chunk[\"gene\"].groupby(chunk['srr'],as_index= True).nunique()\n",
    "        srr_count = len(count_genes)\n",
    "        proc_srr = proc_srr + srr_count       \n",
    "        count_genes = count_genes[count_genes.index.isin(srr_ind)]\n",
    "        local_srr_ind = count_genes[count_genes == n_genes].index\n",
    "        diff = count_genes.index.difference(local_srr_ind)\n",
    "        bad_srrs = [*bad_srrs, *diff]\n",
    "        if len(local_srr_ind):\n",
    "            local_srr_ind = pd.DataFrame(data=range(0, len(local_srr_ind)), index = local_srr_ind, columns = [\"local_pos\"])\n",
    "            chunk = chunk[chunk[\"srr\"].isin(local_srr_ind.index)]\n",
    "            local_srr_ind['global_pos'] = local_srr_ind.index.map(srr_ind)\n",
    "            #row_nums = [local_srr_ind[\"local_pos\"][srr] for srr in chunk[\"srr\"]]\n",
    "            #col_nums = [gene_ind[gene] for gene in chunk[\"gene\"]]\n",
    "            #raw_matrix = csr_matrix((chunk[\"se\"], (row_nums,col_nums)))\n",
    "            raw_matrix = csr_matrix(chunk[\"se\"].values.astype(int).reshape(len(local_srr_ind),n_genes))\n",
    "            A=gsm_from_srr_matrix[:,local_srr_ind[\"global_pos\"]] #size = all_gsm x needed_srr\n",
    "            gsm_mask= np.unique(A.nonzero()[0]) # find needed gsm (rows in matrix above)\n",
    "            A = A.tocsr() \n",
    "            A = A.dot(raw_matrix)\n",
    "            A=A.transpose()\n",
    "            exp_data[:,gsm_mask]+= A[:,gsm_mask] \n",
    "        if proc_srr%500 ==0:\n",
    "            print(proc_srr/true_srr)\n",
    "    print(\"publish meta...\")\n",
    "    print(\"start grouping...\")\n",
    "\n",
    "    for srr in bad_srrs:\n",
    "        for col in col_list:\n",
    "            if (col not in [geo_acc_name, geo_gse_name]):\n",
    "                meta_df.loc[meta_df[\"SRR_accession\"]==srr,col] = \"\"\n",
    "    meta_df =meta_df.groupby(geo_acc_name).agg([lambda x: ';'.join(x)])\n",
    "    print(\"groupped\")\n",
    "    meta_df=meta_df.loc[gsm_ind.keys()]\n",
    "    print(\"reordered\")\n",
    "    gene_info=pd.read_csv(filepath_or_buffer=gene_info_file_name,sep=\"\\t\",index_col =\"GeneID\")\n",
    "    dt = h5py.special_dtype(vlen=np.str)\n",
    "    ensem = np.array(list(gene_ind.keys()),dtype=dt)\n",
    "    gene_symbol = gene_info.loc[ensem][\"GeneSymbol\"]\n",
    "    h5_gse.create_dataset('/meta/genes/ensembl_gene_id',data=ensem,chunks =(len(gene_ind),),compression=\"gzip\", compression_opts=7)\n",
    "    h5_gse.create_dataset('/meta/genes/gene_symbol',data=np.array(gene_symbol, dtype =dt),chunks =(len(gene_ind),),compression=\"gzip\", compression_opts=7)\n",
    "    h5_gse.create_dataset('/meta/info/description', data=\"Contain all SRRs from dee2 for which sample_geo_accession starts with gse. Star estimates was groupped by sample_geo_accession and summed up\")\n",
    "    h5_gse.create_dataset('/meta/genes/genes',data=np.array(gene_symbol, dtype =dt),chunks =(len(gene_ind),),compression=\"gzip\", compression_opts=7)\n",
    "    h5_gse.create_dataset('/meta/samples/geo_accession',data=np.array(list(gsm_ind.keys()), dtype =dt),chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/SRR_accession',data= np.reshape(np.array(meta_df[\"SRR_accession\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/instrument_model',data=np.reshape(np.array(meta_df[\"Model\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/last_update_date',data=np.reshape(np.array(meta_df[\"LoadDate\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/library_selection',data=np.reshape(np.array(meta_df[\"LibrarySelection\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/library_source',data=np.reshape(np.array(meta_df[\"LibrarySource\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/library_strategy',data=np.reshape(np.array(meta_df[\"LibraryStrategy\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/organism_ch1',data=np.reshape(np.array(meta_df[\"ScientificName\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/series_id',data=np.reshape(np.array(meta_df[geo_gse_name]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/status',data=np.reshape(np.array(meta_df[\"Consent\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/submission_date',data=np.reshape(np.array(meta_df[\"ReleaseDate\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/quality',data=np.reshape(np.array(meta_df[\"QC_summary\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/title',data=np.reshape(np.array(meta_df[\"BioSample\"]),-1),dtype=dt,chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    h5_gse.create_dataset('/meta/samples/type',data=np.full(len(meta_df[\"BioSample\"]),\"SRA\", dtype=dt),chunks =(2000,),compression=\"gzip\", compression_opts=5)\n",
    "    \n",
    "    h5_gse.flush()\n",
    "print(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "make gene index...\n",
      "gene quantity: 58302\n",
      "read meta file...\n",
      "Wall time: 16 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import datetime\n",
    "import bz2\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.sparse as sps\n",
    "from scipy.sparse import isspmatrix_csr,csr_matrix,csc_matrix,issparse,isspmatrix_csr, lil_matrix\n",
    "import math\n",
    "cache_dir=\"G://datasets/DEE2/\"\n",
    "#cache_dir=\"./\"\n",
    "organism_name= 'hsapiens'#'hsapiens' 'ecoli' 'mmusculus' 'athaliana' 'drerio' 'scerevisiae'celegans dmelanogaster  rnorvegicus\n",
    "metadata_date = \"20210319\"\n",
    "metadata_file = cache_dir+\"metadata/\"+ organism_name+\"_metadata_\"+metadata_date+\".tsv\"\n",
    "source_date = \"20210319\"\n",
    "source_name= organism_name+\"_se_\"+source_date+\".tsv.bz2\"\n",
    "data_file_name= cache_dir+\"raw_files/star_counts/\"+source_name #celegans_se_20200530.tsv.bz2 dmelanogaster_se_20200530.tsv.bz2 ecoli_se_20200530.tsv.bz2\n",
    "#drerio_se_20200530.tsv.bz2\n",
    "#hsapiens_se_20200528.tsv.bz2\n",
    "h5_gsm_name=cache_dir+\"star_h5/\"+organism_name+ \"_star_matrix_\"+source_date+\".h5\"\n",
    "#h5_gsm_name = \"/scratch2/mkleverov/dee2/star_h5/\"+organism_name+ \"_star_matrix.h5\"\n",
    "gene_info_file_name =cache_dir+\"GeneInfo/\"+organism_name+\"_GeneInfo.tsv\"\n",
    "dt = h5py.special_dtype(vlen=np.str)\n",
    "now = datetime.datetime.now()\n",
    "print(\"make gene index...\")\n",
    "gene_info=pd.read_csv(filepath_or_buffer=gene_info_file_name,sep=\"\\t\")\n",
    "gene_ind= {gene:ind for ind,gene in enumerate(gene_info[\"GeneID\"])}\n",
    "n_genes=len(gene_ind)\n",
    "print(\"gene quantity: \"+str(n_genes))\n",
    "print(\"read meta file...\")\n",
    "iter_meta = pd.read_csv(filepath_or_buffer=metadata_file,sep=\"\\t\", iterator=True, chunksize=1000)\n",
    "meta_df = pd.concat([chunk for chunk in iter_meta])\n",
    "geo_acc_name = meta_df.columns[5]#'GSE_accession' # may be \"GSE_accession\" or \"Sample_name\"\n",
    "geo_gse_name = meta_df.columns[6]#'Sample_name' # may be \"GEO_series\" \"SampleName\" or \"Sample_name\"(ecoli)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR_accession</th>\n",
       "      <th>QC_summary</th>\n",
       "      <th>SRX_accession</th>\n",
       "      <th>SRS_accession</th>\n",
       "      <th>SRP_accession</th>\n",
       "      <th>Sample_name</th>\n",
       "      <th>GEO_series</th>\n",
       "      <th>Library_name</th>\n",
       "      <th>SampleName</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>...</th>\n",
       "      <th>Affection_Status</th>\n",
       "      <th>Analyte_Type</th>\n",
       "      <th>Histological_Type</th>\n",
       "      <th>Body_Site</th>\n",
       "      <th>CenterName</th>\n",
       "      <th>Submission</th>\n",
       "      <th>dbgap_study_accession</th>\n",
       "      <th>Consent</th>\n",
       "      <th>RunHash</th>\n",
       "      <th>ReadHash</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DRR000003</td>\n",
       "      <td>FAIL(4,5,6,7)</td>\n",
       "      <td>DRX000003</td>\n",
       "      <td>DRS000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DLD1_normoxia_nucleosome</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>2010-10-14 00:53:29</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>57A45899FA02E1A183A0ACA9F65E6A0A</td>\n",
       "      <td>D602C0CD8F2B562BA085FB8FC49C964B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DRR000004</td>\n",
       "      <td>FAIL(4,5,6,7)</td>\n",
       "      <td>DRX000003</td>\n",
       "      <td>DRS000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DLD1_normoxia_nucleosome</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>2010-10-14 00:53:29</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>D246572C1CAADBF3F601A23B6F7CD289</td>\n",
       "      <td>6FC67E696F9658B211BD15844D89C4AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DRR000005</td>\n",
       "      <td>FAIL(4,5,6,7)</td>\n",
       "      <td>DRX000003</td>\n",
       "      <td>DRS000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DLD1_normoxia_nucleosome</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>2010-10-14 00:53:29</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>92FC08D16F287F47B6AE70F92CD0AB4C</td>\n",
       "      <td>025670D28DB6462F558A4C31DC4D646B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DRR000006</td>\n",
       "      <td>FAIL(4,5,6,7)</td>\n",
       "      <td>DRX000003</td>\n",
       "      <td>DRS000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DLD1_normoxia_nucleosome</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>2010-10-14 00:53:29</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>E9B011701C2E7B972FAF33CD29840F53</td>\n",
       "      <td>D161FC28BB386893EA52829DFF3D16C7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DRR000007</td>\n",
       "      <td>FAIL(4,5,6,7)</td>\n",
       "      <td>DRX000003</td>\n",
       "      <td>DRS000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DLD1_normoxia_nucleosome</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>2010-10-14 00:53:29</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>850599A62FFA87DF2E39281FC127C397</td>\n",
       "      <td>BDD124DC6EFF61D7A50A9CA6F8A8F6EB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604971</th>\n",
       "      <td>SRR9995485</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SRX6734453</td>\n",
       "      <td>SRS5285729</td>\n",
       "      <td>SRP218732</td>\n",
       "      <td>GSM4038638</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4038638</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRA944214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>1BAD74B91D3650838330DF8834ECFF93</td>\n",
       "      <td>E23A78A8A16CCF9D3C76BE63D0FAB6B0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604972</th>\n",
       "      <td>SRR9995486</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SRX6734453</td>\n",
       "      <td>SRS5285729</td>\n",
       "      <td>SRP218732</td>\n",
       "      <td>GSM4038638</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4038638</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRA944214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>601BBF0AEA492785E30D9997022B628C</td>\n",
       "      <td>E58BD173F8E2A8A88D9D0B9BE45666AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604973</th>\n",
       "      <td>SRR9995490</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SRX6734455</td>\n",
       "      <td>SRS5285731</td>\n",
       "      <td>SRP218732</td>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRA944214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>70753910DD0C02C64FD6541AC2028D0F</td>\n",
       "      <td>44662CBEBBC66B6DFBB06A20B74685AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604974</th>\n",
       "      <td>SRR9995491</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SRX6734455</td>\n",
       "      <td>SRS5285731</td>\n",
       "      <td>SRP218732</td>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRA944214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>13BEC07E0A6B45D76FF4F181590E252B</td>\n",
       "      <td>E625B95FE9A20A7CB4742F8C671F3C4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604975</th>\n",
       "      <td>SRR9995492</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SRX6734455</td>\n",
       "      <td>SRS5285731</td>\n",
       "      <td>SRP218732</td>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRA944214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>public</td>\n",
       "      <td>CD062B887D5DFEB6E8FB144C2FEA9995</td>\n",
       "      <td>FBC4CCD5FC664F1E3100836A78FFAD46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>604976 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       SRR_accession     QC_summary SRX_accession SRS_accession SRP_accession  \\\n",
       "0          DRR000003  FAIL(4,5,6,7)     DRX000003     DRS000003     DRP000003   \n",
       "1          DRR000004  FAIL(4,5,6,7)     DRX000003     DRS000003     DRP000003   \n",
       "2          DRR000005  FAIL(4,5,6,7)     DRX000003     DRS000003     DRP000003   \n",
       "3          DRR000006  FAIL(4,5,6,7)     DRX000003     DRS000003     DRP000003   \n",
       "4          DRR000007  FAIL(4,5,6,7)     DRX000003     DRS000003     DRP000003   \n",
       "...              ...            ...           ...           ...           ...   \n",
       "604971    SRR9995485    WARN(1,5,7)    SRX6734453    SRS5285729     SRP218732   \n",
       "604972    SRR9995486    WARN(1,5,7)    SRX6734453    SRS5285729     SRP218732   \n",
       "604973    SRR9995490    WARN(1,5,7)    SRX6734455    SRS5285731     SRP218732   \n",
       "604974    SRR9995491    WARN(1,5,7)    SRX6734455    SRS5285731     SRP218732   \n",
       "604975    SRR9995492    WARN(1,5,7)    SRX6734455    SRS5285731     SRP218732   \n",
       "\n",
       "         Sample_name  GEO_series              Library_name    SampleName  \\\n",
       "0       SAMD00013986         NaN  DLD1_normoxia_nucleosome  SAMD00013986   \n",
       "1       SAMD00013986         NaN  DLD1_normoxia_nucleosome  SAMD00013986   \n",
       "2       SAMD00013986         NaN  DLD1_normoxia_nucleosome  SAMD00013986   \n",
       "3       SAMD00013986         NaN  DLD1_normoxia_nucleosome  SAMD00013986   \n",
       "4       SAMD00013986         NaN  DLD1_normoxia_nucleosome  SAMD00013986   \n",
       "...              ...         ...                       ...           ...   \n",
       "604971    GSM4038638         NaN                       NaN    GSM4038638   \n",
       "604972    GSM4038638         NaN                       NaN    GSM4038638   \n",
       "604973    GSM4038640         NaN                       NaN    GSM4038640   \n",
       "604974    GSM4038640         NaN                       NaN    GSM4038640   \n",
       "604975    GSM4038640         NaN                       NaN    GSM4038640   \n",
       "\n",
       "                ReleaseDate  ... Affection_Status  Analyte_Type  \\\n",
       "0       2010-10-14 00:53:29  ...              NaN           NaN   \n",
       "1       2010-10-14 00:53:29  ...              NaN           NaN   \n",
       "2       2010-10-14 00:53:29  ...              NaN           NaN   \n",
       "3       2010-10-14 00:53:29  ...              NaN           NaN   \n",
       "4       2010-10-14 00:53:29  ...              NaN           NaN   \n",
       "...                     ...  ...              ...           ...   \n",
       "604971  2020-06-12 16:15:18  ...              NaN           NaN   \n",
       "604972  2020-06-12 16:15:18  ...              NaN           NaN   \n",
       "604973  2020-06-12 16:15:18  ...              NaN           NaN   \n",
       "604974  2020-06-12 16:15:18  ...              NaN           NaN   \n",
       "604975  2020-06-12 16:15:18  ...              NaN           NaN   \n",
       "\n",
       "        Histological_Type  Body_Site  CenterName  Submission  \\\n",
       "0                     NaN        NaN      UT-MGS   DRA000003   \n",
       "1                     NaN        NaN      UT-MGS   DRA000003   \n",
       "2                     NaN        NaN      UT-MGS   DRA000003   \n",
       "3                     NaN        NaN      UT-MGS   DRA000003   \n",
       "4                     NaN        NaN      UT-MGS   DRA000003   \n",
       "...                   ...        ...         ...         ...   \n",
       "604971                NaN        NaN         GEO   SRA944214   \n",
       "604972                NaN        NaN         GEO   SRA944214   \n",
       "604973                NaN        NaN         GEO   SRA944214   \n",
       "604974                NaN        NaN         GEO   SRA944214   \n",
       "604975                NaN        NaN         GEO   SRA944214   \n",
       "\n",
       "       dbgap_study_accession Consent                           RunHash  \\\n",
       "0                        NaN  public  57A45899FA02E1A183A0ACA9F65E6A0A   \n",
       "1                        NaN  public  D246572C1CAADBF3F601A23B6F7CD289   \n",
       "2                        NaN  public  92FC08D16F287F47B6AE70F92CD0AB4C   \n",
       "3                        NaN  public  E9B011701C2E7B972FAF33CD29840F53   \n",
       "4                        NaN  public  850599A62FFA87DF2E39281FC127C397   \n",
       "...                      ...     ...                               ...   \n",
       "604971                   NaN  public  1BAD74B91D3650838330DF8834ECFF93   \n",
       "604972                   NaN  public  601BBF0AEA492785E30D9997022B628C   \n",
       "604973                   NaN  public  70753910DD0C02C64FD6541AC2028D0F   \n",
       "604974                   NaN  public  13BEC07E0A6B45D76FF4F181590E252B   \n",
       "604975                   NaN  public  CD062B887D5DFEB6E8FB144C2FEA9995   \n",
       "\n",
       "                                ReadHash  \n",
       "0       D602C0CD8F2B562BA085FB8FC49C964B  \n",
       "1       6FC67E696F9658B211BD15844D89C4AC  \n",
       "2       025670D28DB6462F558A4C31DC4D646B  \n",
       "3       D161FC28BB386893EA52829DFF3D16C7  \n",
       "4       BDD124DC6EFF61D7A50A9CA6F8A8F6EB  \n",
       "...                                  ...  \n",
       "604971  E23A78A8A16CCF9D3C76BE63D0FAB6B0  \n",
       "604972  E58BD173F8E2A8A88D9D0B9BE45666AD  \n",
       "604973  44662CBEBBC66B6DFBB06A20B74685AD  \n",
       "604974  E625B95FE9A20A7CB4742F8C671F3C4A  \n",
       "604975  FBC4CCD5FC664F1E3100836A78FFAD46  \n",
       "\n",
       "[604976 rows x 53 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_name</th>\n",
       "      <th>SRR_accession</th>\n",
       "      <th>Model</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>GEO_series</th>\n",
       "      <th>Consent</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>QC_summary</th>\n",
       "      <th>BioSample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126715</th>\n",
       "      <td>GSM402329</td>\n",
       "      <td>SRR029054</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "      <td>2014-05-26 09:27:28</td>\n",
       "      <td>other</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2009-10-21 17:33:21</td>\n",
       "      <td>FAIL(3,4,5,6,7)</td>\n",
       "      <td>SAMN00004353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126716</th>\n",
       "      <td>GSM416753</td>\n",
       "      <td>SRR029124</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "      <td>2014-05-26 09:26:27</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2009-11-03 13:20:36</td>\n",
       "      <td>FAIL(3,4,5,6,7)</td>\n",
       "      <td>SAMN00004408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126717</th>\n",
       "      <td>GSM416754</td>\n",
       "      <td>SRR029125</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "      <td>2014-05-26 09:26:29</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2009-11-03 13:20:48</td>\n",
       "      <td>FAIL(3,4,5,6,7)</td>\n",
       "      <td>SAMN00004409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126718</th>\n",
       "      <td>GSM416755</td>\n",
       "      <td>SRR029126</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "      <td>2014-05-26 09:26:31</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2009-11-03 13:20:31</td>\n",
       "      <td>FAIL(3,4,5,6,7)</td>\n",
       "      <td>SAMN00004410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126719</th>\n",
       "      <td>GSM416756</td>\n",
       "      <td>SRR029127</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "      <td>2014-05-26 09:26:35</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2009-11-03 13:20:57</td>\n",
       "      <td>FAIL(3,4,5,6,7)</td>\n",
       "      <td>SAMN00004411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604971</th>\n",
       "      <td>GSM4038638</td>\n",
       "      <td>SRR9995485</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>2019-08-19 12:14:08</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SAMN12600296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604972</th>\n",
       "      <td>GSM4038638</td>\n",
       "      <td>SRR9995486</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>2019-08-19 12:16:54</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SAMN12600296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604973</th>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>SRR9995490</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>2019-08-19 12:17:36</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SAMN12600294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604974</th>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>SRR9995491</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>2019-08-19 12:15:20</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SAMN12600294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604975</th>\n",
       "      <td>GSM4038640</td>\n",
       "      <td>SRR9995492</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>2019-08-19 12:15:30</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2020-06-12 16:15:18</td>\n",
       "      <td>WARN(1,5,7)</td>\n",
       "      <td>SAMN12600294</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>376153 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample_name SRR_accession                     Model  \\\n",
       "126715   GSM402329     SRR029054  Illumina Genome Analyzer   \n",
       "126716   GSM416753     SRR029124  Illumina Genome Analyzer   \n",
       "126717   GSM416754     SRR029125  Illumina Genome Analyzer   \n",
       "126718   GSM416755     SRR029126  Illumina Genome Analyzer   \n",
       "126719   GSM416756     SRR029127  Illumina Genome Analyzer   \n",
       "...            ...           ...                       ...   \n",
       "604971  GSM4038638    SRR9995485       Illumina HiSeq 2500   \n",
       "604972  GSM4038638    SRR9995486       Illumina HiSeq 2500   \n",
       "604973  GSM4038640    SRR9995490       Illumina HiSeq 2500   \n",
       "604974  GSM4038640    SRR9995491       Illumina HiSeq 2500   \n",
       "604975  GSM4038640    SRR9995492       Illumina HiSeq 2500   \n",
       "\n",
       "                   LoadDate    LibrarySelection   LibrarySource  \\\n",
       "126715  2014-05-26 09:27:28               other  TRANSCRIPTOMIC   \n",
       "126716  2014-05-26 09:26:27  size fractionation  TRANSCRIPTOMIC   \n",
       "126717  2014-05-26 09:26:29  size fractionation  TRANSCRIPTOMIC   \n",
       "126718  2014-05-26 09:26:31  size fractionation  TRANSCRIPTOMIC   \n",
       "126719  2014-05-26 09:26:35  size fractionation  TRANSCRIPTOMIC   \n",
       "...                     ...                 ...             ...   \n",
       "604971  2019-08-19 12:14:08                cDNA  TRANSCRIPTOMIC   \n",
       "604972  2019-08-19 12:16:54                cDNA  TRANSCRIPTOMIC   \n",
       "604973  2019-08-19 12:17:36                cDNA  TRANSCRIPTOMIC   \n",
       "604974  2019-08-19 12:15:20                cDNA  TRANSCRIPTOMIC   \n",
       "604975  2019-08-19 12:15:30                cDNA  TRANSCRIPTOMIC   \n",
       "\n",
       "       LibraryStrategy ScientificName GEO_series Consent          ReleaseDate  \\\n",
       "126715           OTHER   Homo sapiens        nan  public  2009-10-21 17:33:21   \n",
       "126716         RNA-Seq   Homo sapiens        nan  public  2009-11-03 13:20:36   \n",
       "126717         RNA-Seq   Homo sapiens        nan  public  2009-11-03 13:20:48   \n",
       "126718         RNA-Seq   Homo sapiens        nan  public  2009-11-03 13:20:31   \n",
       "126719         RNA-Seq   Homo sapiens        nan  public  2009-11-03 13:20:57   \n",
       "...                ...            ...        ...     ...                  ...   \n",
       "604971         RNA-Seq   Homo sapiens        nan  public  2020-06-12 16:15:18   \n",
       "604972         RNA-Seq   Homo sapiens        nan  public  2020-06-12 16:15:18   \n",
       "604973         RNA-Seq   Homo sapiens        nan  public  2020-06-12 16:15:18   \n",
       "604974         RNA-Seq   Homo sapiens        nan  public  2020-06-12 16:15:18   \n",
       "604975         RNA-Seq   Homo sapiens        nan  public  2020-06-12 16:15:18   \n",
       "\n",
       "             QC_summary     BioSample  \n",
       "126715  FAIL(3,4,5,6,7)  SAMN00004353  \n",
       "126716  FAIL(3,4,5,6,7)  SAMN00004408  \n",
       "126717  FAIL(3,4,5,6,7)  SAMN00004409  \n",
       "126718  FAIL(3,4,5,6,7)  SAMN00004410  \n",
       "126719  FAIL(3,4,5,6,7)  SAMN00004411  \n",
       "...                 ...           ...  \n",
       "604971      WARN(1,5,7)  SAMN12600296  \n",
       "604972      WARN(1,5,7)  SAMN12600296  \n",
       "604973      WARN(1,5,7)  SAMN12600294  \n",
       "604974      WARN(1,5,7)  SAMN12600294  \n",
       "604975      WARN(1,5,7)  SAMN12600294  \n",
       "\n",
       "[376153 rows x 13 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_srrs = ['SRR029054','SRR9995486']\n",
    "for srr in bad_srrs:\n",
    "        ind = meta_df[\"SRR_accession\"]==srr\n",
    "        for col in col_list:\n",
    "            if (col not in [geo_acc_name, geo_gse_name]):     \n",
    "                meta_df.loc[ind,col] = \"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_srrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_name</th>\n",
       "      <th>SRR_accession</th>\n",
       "      <th>Model</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>GEO_series</th>\n",
       "      <th>Consent</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>QC_summary</th>\n",
       "      <th>BioSample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Sample_name, SRR_accession, Model, LoadDate, LibrarySelection, LibrarySource, LibraryStrategy, ScientificName, GEO_series, Consent, ReleaseDate, QC_summary, BioSample]\n",
       "Index: []"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.loc[meta_df[\"SRR_accession\"]==\"SRR029054\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df.loc[meta_df[\"SRR_accession\"]==\"SRR029124\",'LoadDate'] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df.loc[meta_df[\"SRR_accession\"]==\"SRR029124\", [\"LibrarySelection\", \"ScientificName\"] ] =\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_name</th>\n",
       "      <th>SRR_accession</th>\n",
       "      <th>Model</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>ScientificName</th>\n",
       "      <th>GEO_series</th>\n",
       "      <th>Consent</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>QC_summary</th>\n",
       "      <th>BioSample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126716</th>\n",
       "      <td>GSM416753</td>\n",
       "      <td>SRR029124</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td></td>\n",
       "      <td>nan</td>\n",
       "      <td>public</td>\n",
       "      <td>2009-11-03 13:20:36</td>\n",
       "      <td>FAIL(3,4,5,6,7)</td>\n",
       "      <td>SAMN00004408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample_name SRR_accession                     Model LoadDate  \\\n",
       "126716   GSM416753     SRR029124  Illumina Genome Analyzer            \n",
       "\n",
       "       LibrarySelection   LibrarySource LibraryStrategy ScientificName  \\\n",
       "126716                   TRANSCRIPTOMIC         RNA-Seq                  \n",
       "\n",
       "       GEO_series Consent          ReleaseDate       QC_summary     BioSample  \n",
       "126716        nan  public  2009-11-03 13:20:36  FAIL(3,4,5,6,7)  SAMN00004408  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.loc[meta_df[\"SRR_accession\"]==\"SRR029124\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "126716    \n",
       "Name: LibrarySelection, dtype: object"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.loc[meta_df[\"SRR_accession\"]==\"SRR029124\", \"LibrarySelection\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1, 0, 2, 3, 1]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[2, 1, 0, 0],\n",
       "       [0, 1, 1, 1]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csr_matrix((data, indices, indptr), dtype=int).toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 3, 1]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indices[indptr[1]:indptr[2]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SRR/h5 file with matrices Работает так же долго"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start h5\n",
      "4322\n",
      "start cast matrices\n",
      "start process expression\n",
      "0.14236902050113895\n",
      "0.2847380410022779\n",
      "0.4271070615034169\n",
      "0.5694760820045558\n",
      "0.7118451025056948\n",
      "0.8542141230068337\n",
      "0.9965831435079726\n",
      "1\n",
      "Wall time: 29 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "import datetime\n",
    "from scipy.sparse import isspmatrix_csr,csr_matrix,csc_matrix\n",
    "data_file_name= cache_dir+\"raw_files/\"+organism_name+\"_ke.tsv.bz2\"\n",
    "#data_file_name=\"F://mmusculus_ke.tsv.bz2\"\n",
    "h5_gsm_name = cache_dir+organism_name+ \"_kalisto_SRR_matrix.h5\"\n",
    "now = datetime.datetime.now()\n",
    "print(\"start h5\")\n",
    "with h5py.File(h5_gsm_name, 'w') as h5_gse:\n",
    "    meta_grp = h5_gse.create_group('meta')\n",
    "    info_grp = h5_gse.create_group('info')\n",
    "    data_grp= h5_gse.create_group('data')\n",
    "    info_grp.create_dataset('version', data=\"dee2_gse_v1\")\n",
    "    info_grp.create_dataset('creation_date', data=now.strftime(\"%Y.%m.%d\"))\n",
    "    n_genes =len(gene_ind)\n",
    "    dt = h5py.special_dtype(vlen=np.str) \n",
    "    meta_grp.create_dataset('genes',data=srr_to_gsm.values(),dtype=dt)\n",
    "    meta_grp.create_dataset('SRR',data=srr_ind.keys(),dtype=dt)\n",
    "    meta_grp.create_dataset('Sample_geo_accession',data=meta_df[geo_acc_name], dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_instrument_model',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_last_update_date',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_library_selection',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_library_source',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_library_strategy',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_organism_ch1',(n_srr,),dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_series_id',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_status',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_submission_date',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_taxid_ch1',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_quality',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_title',(n_srr,), dtype =dt)\n",
    "    meta_grp.create_dataset('Sample_type',data=np.full(n_srr,\"SRA\", dtype =dt))\n",
    "    exp_data=data_grp.create_dataset(\"expression\", (n_srr,n_genes),dtype= 'i4')#, compression=\"gzip\", compression_opts=9)\n",
    "    srr_per_time=1000\n",
    "    row_length= len(tx_ind)\n",
    "    print(row_length)\n",
    "    iter_data = pd.read_csv(filepath_or_buffer=data_file_name,sep=\"\\t\", iterator=True, chunksize=srr_per_time*row_length,names=[\"srr\",\"tx\",\"ke\"])\n",
    "    proc_srr= 0\n",
    "    print(\"start cast matrices\")  \n",
    "    tx_to_gene_matrix=csr_matrix(tx_to_gene_matrix)\n",
    "    print(\"start process expression\")\n",
    "    for chunk in iter_data:\n",
    "        global_ids =[]\n",
    "        local_ids=[]\n",
    "        local_srr_ind= {srr:ind for ind,srr in enumerate(pd.unique(chunk['srr']))}\n",
    "        srr_count=len(local_srr_ind)\n",
    "        proc_srr=proc_srr+srr_count\n",
    "        for cur_srr in sorted(local_srr_ind.keys()):\n",
    "            try:\n",
    "                srr_id= srr_ind[cur_srr]\n",
    "                global_ids.append(srr_id)\n",
    "                local_ids.append(local_srr_ind[cur_srr])\n",
    "            except(KeyError):\n",
    "                continue\n",
    "        \n",
    "        if local_ids:           \n",
    "            raw_matrix= chunk[\"ke\"].values.astype(int).reshape(srr_count,row_length)\n",
    "            A=csr_matrix(raw_matrix[local_ids,:])\n",
    "            A = A.dot(tx_to_gene_matrix)\n",
    "            exp_data[global_ids,0:n_genes]= exp_data[global_ids,0:n_genes]+A[:,0:n_genes]\n",
    "        if proc_srr%500 ==0:\n",
    "            print(proc_srr/true_srr)\n",
    "    h5_gse.flush()\n",
    "print(1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "dee2 kalisto to h5 files"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
